src/vs/workbench/contrib/terminalContrib/suggest/browser/terminalCompletionService.ts

```diff
@@ -110,31 +148,92 @@ export class TerminalCompletionService extends Disposable implements ITerminalCo
 			providers = providers.filter(p => p.isBuiltin);
 		}
 
-		await this._collectCompletions(providers, shellType, promptValue, cursorPosition, completionItems, token);
-		return completionItems.length > 0 ? completionItems : undefined;
+		return this._collectCompletions(providers, shellType, promptValue, cursorPosition, token);
 	}
 
-	private async _collectCompletions(providers: ITerminalCompletionProvider[], shellType: TerminalShellType, promptValue: string, cursorPosition: number, completionItems: ISimpleCompletion[], token: CancellationToken) {
+	private async _collectCompletions(providers: ITerminalCompletionProvider[], shellType: TerminalShellType, promptValue: string, cursorPosition: number, token: CancellationToken): Promise<ITerminalCompletion[] | undefined> {
 		const completionPromises = providers.map(async provider => {
 			if (provider.shellTypes && !provider.shellTypes.includes(shellType)) {
-				return [];
+				return undefined;
+			}
+			const completions: ITerminalCompletion[] | TerminalCompletionList<ITerminalCompletion> | undefined = await provider.provideCompletions(promptValue, cursorPosition, token);
+			if (!completions) {
+				return undefined;
 			}
-			const completions = await provider.provideCompletions(promptValue, cursorPosition, token);
 			const devModeEnabled = this._configurationService.getValue(TerminalSettingId.DevMode);
-			if (completions) {
-				return completions.map(completion => {
-					if (devModeEnabled && !completion.detail?.includes(provider.id)) {
-						completion.detail = `(${provider.id}) ${completion.detail ?? ''}`;
-					}
-					return completion;
-				});
+			const completionItems = Array.isArray(completions) ? completions : completions.items ?? [];
```

Comment:
:+1:

Ah, I was almost going to leave another suggestion to make a helper function `isCompletionsList` which returns a type like `value is TerminalCompletionsList` to help with the assertion. Would lead to more readable code.